src/fs/metadata_log/entries.hh

```diff
@@ -0,0 +1,255 @@
+/*
+ * This file is open source software, licensed to you under the terms
+ * of the Apache License, Version 2.0 (the "License").  See the NOTICE file
+ * distributed with this work for additional information regarding copyright
+ * ownership.  You may not use this file except in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*
+ * Copyright (C) 2020 ScyllaDB
+ */
+
+#pragma once
+
+#include "seastar/core/temporary_buffer.hh"
+#include "fs/cluster.hh"
+#include "fs/inode.hh"
+#include "fs/units.hh"
+#include "fs/unix_metadata.hh"
+#include "fs/val_or_err.hh"
+
+#include <string>
+#include <string_view>
+
+namespace seastar::fs::metadata_log::entries {
+
+struct checkpoint {
+    // The disk format is as follows:
+    // | checkpoint | .............................. |
+    //              |             data               |
+    //              |<-- checkpointed_data_length -->|
+    //                                               ^
+    //       _______________________________________/
+    //      /
+    //    there ends checkpointed data and (next checkpoint begins or metadata in the current cluster end)
+    //
+    // CRC is calculated from byte sequence | data | checkpointed_data_length |
+    // E.g. if the data consist of bytes "abcd" and checkpointed_data_length of bytes "xyz" then the byte sequence
+    // would be "abcdxyz"
+    uint32_t crc32_checksum;
+    unit_size_t checkpointed_data_length;
+};
+
+struct next_metadata_cluster {
+    cluster_id_t cluster_id; // metadata log continues there
+};
+
+struct create_inode {
+    inode_t inode;
+    unix_metadata metadata;
+};
+
+struct delete_inode {
+    inode_t inode;
+};
+
+struct small_write {
+    inode_t inode;
+    file_offset_t offset;
+    decltype(unix_metadata::mtime_ns) time_ns;
```

Comment:
just place `using mtime_ns = unix_metadata::mtime_ns` inside this namespace to avoid boilerplate decltypes - applies to all uses of `decltype` in this file.

Reply:
Maybe `using time_ns_t = ...` as it is uint64_t underneath?